sql: teach instanceResolver to (strictly) match multiple on localities

msbutler opened the pull request:
This patch teaches distSQL planning's instanceResolver (maps nodes to sql
instances) to match on multiple localities instead of just one. Further, if the
distsql planner is intialized with strict filtering, the distsql plan will
return an error if sql instance returned does not match the passed in
localities.

This work will enable data domiciled locality aware backups. For example,
suppose the user wants to back up an MR cluster spread over regions X, Y and Z.
With strict filtering, the user must pass 3 URIs each with a locality tag to X,
Y, Z. These locality pairs will then get plumbed to dist sql planning to ensure
that data stored in region X will get processed by a node in region X.
Downstream logic ensures a node in region X will send data to the bucket in
region X. So, suppose the user did not pass a bucket with locality Z, then
distsql planning would fail, as it could not return an instance that matches
the filters [X,Y] which contains data in Z.

Informs

Release note: none

msbutler committed:
kvfollowerreadsccl: allow several localities in bulk oracle

With this patch, the bulk oracle can now match on several localities instead
of one.

This patch also splits up NewBulkOracle into NewStreakBulkOracle and
NewLocalityFilteringBulkOracle, because filtering and streaking cannot occur at
the same time, tracked in #120755.

This work is apart of a larger project to enable BACKUP with EXECUTION LOCALITY
with multiple localities and data domiciled locality aware backups.

Informs #155788

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/156474)
<!-- Reviewable:end -->

msbutler commented on pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:
```diff
@@ -267,17 +276,22 @@ func (r bulkOracle) ChoosePreferredReplica(
 		// the code below to pick one as normal, just from the filtered slice.
 		if len(matches) > 0 {
 			return replicas[matches[randutil.FastUint32()%uint32(len(matches))]].ReplicaDescriptor, true, nil
+		} else if r.strictFiltering {
```
hrm, now that i spell this out to you, i think my current implementation is insufficient.

essentially, i'd like to guarantee the following: if a range's replicas are _constrained_ to region X, i want the bulk oracle logic to guarantee that the range's span will get processed by a node in region X. If no node is available in region X, return an error.  if we match the range to the right node, we can assume the backup processor will back up the data to the right uri (perhaps we can add assertions in backup processor for this too).

I think what's wrong with my current implementation is that it's not relying on zone configs, rather it's relying on current replica placement. In other words, if the allocator is not complying with the persisted zone configs, my backup could still succeed. Let me see if there's some `GetZoneConfig(rangeID)` api.

dt replied:
I’m not sure?

zone configs are what control the placement of replicas and seems a little like the wrong layer/violation of separation of concerns to check them here.

I thought our current behavior of saying data should be backed up where it resides, ie where the replicas are, was pretty defensible _if we could promise to actually do that_ but that last part is where sql planning’s best effort means backup can skew from replica placement: my impression is our issue is more with the replica to instance resolution step than the replica resolution step?

msbutler committed:
sql: teach instanceResolver to (strictly) match multiple on localities

This patch teaches distSQL planning's instanceResolver (maps nodes to sql
instances) to match on multiple localities instead of just one. Further, if the
distsql planner is intialized with strict filtering, the distsql plan will
return an error if sql instance returned does not match the passed in
localities.

This work will enable data domiciled locality aware backups. For example,
suppose the user wants to back up an MR cluster spread over regions X, Y and Z.
With strict filtering, the user must pass 3 URIs each with a locality tag to X,
Y, Z. These locality pairs will then get plumbed to dist sql planning to ensure
that data stored in region X will get processed by a node in region X.
Downstream logic ensures a node in region X will send data to the bucket in
region X. So, suppose the user did not pass a bucket with locality Z, then
distsql planning would fail, as it could not return an instance that matches
the filters [X,Y] which contains data in Z.

Informs

Release note: none

msbutler committed:
sql/dsp: add SingleLocalityFilter utility

Epic: none

Release note: none

yuzefovich reviewed (commented):
The change looks good to me, I only have some nits in tests, but I'll defer to David for approval.

> @yuzefovich reviewed 1 of 6 files at r2, 10 of 10 files at r5, 5 of 5 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156474#-:-OfXZojX96pxBAgTziQa:b-n55h3l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kev-cao, @msbutler, and @mw5h)

___
*[`pkg/sql/distsql_physical_planner_test.go` line 928 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/156474#-OfXOUAmD_Yn-cxsWj9s:-OfXOUAmD_Yn-cxsWj9t:b-oy5bez) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3719b51051b3cf8276eb24f4d2ae0b0d37db3cd/pkg/sql/distsql_physical_planner_test.go#L928)):*
> ```Go
> 
> 	localityTag := func(nodeID int) string {
> 		// If you don't have cpu implant: 1: "x=1,y=1", 2: "x=1,y=0", 3: "x=2,y=1"
> ```

nit: I don't understand the beginning of this comment - can you ELI5 :smiley: 
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1258 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/156474#-OfXQ9S1EtO2SNxwu5ju:-OfXQ9S1EtO2SNxwu5jv:bbawav1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3719b51051b3cf8276eb24f4d2ae0b0d37db3cd/pkg/sql/distsql_physical_planner_test.go#L1258)):*
> ```Go
> 
> 			spans:     [][2]string{{"A1", "C1"}, {"D1", "X"}},
> 			locFilter: []string{"x=1"}, // nodes 1, 2, and 3 should match this.
> ```

nit: n3 doesn't match, right?
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1383 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/156474#-OfXOoM38ia1SbNlS4Aw:-OfXOoM38ia1SbNlS4Ax:b-c54p91) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3719b51051b3cf8276eb24f4d2ae0b0d37db3cd/pkg/sql/distsql_physical_planner_test.go#L1383)):*
> ```Go
> 		},
> 
> 		// 13: filter with multiple localities. esentially the same as test case 9,
> ```

nit: `s/esentially/Essentially/`.
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1652 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/156474#-OfXRSTNApeuIFFBsbAJ:-OfXRSTOAyb50AJ30h_3:b-eyrsru) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3719b51051b3cf8276eb24f4d2ae0b0d37db3cd/pkg/sql/distsql_physical_planner_test.go#L1652)):*
> ```Go
> 				planCtxStrict := dsp.NewPlanningCtxWithOracle(
> 					ctx, &extendedEvalContext{Context: *evalCtx}, nil, /* planner */
> 					nil /* txn */, FullDistribution, physicalplan.DefaultReplicaChooser, locFilters, true,
> ```

nit: add inlined comment for `true`.


<!-- Sent from Reviewable.io -->

dt commented on pkg/roachpb/metadata.go:
```diff
@@ -650,6 +650,13 @@ func (l Locality) Matches(filter Locality) (bool, Tier) {
 	return true, Tier{}
 }
 
+func (l Locality) NonEmptySlice() []Locality {
```
nit: it too me a minute to figure out what this was doing -- I'm not sure why but I find this helper really unintuitive for some reason (maybe since a locality is a slice of tiers itself). what about moving it to pkg/sql/dsp.go  as `func SingleLocalityFilter(Locality) []Locality { <same body> }`?

msbutler commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -1935,13 +1964,41 @@ func (dsp *DistSQLPlanner) makeInstanceResolver(
+				if !matched {
+					// We could still call CloseInstances, but in strict mode, the caller
+					// would still need to error, so there is no point.
+					return instances[rng.Intn(len(instances))].InstanceID, SpanPartitionReason_LOCALITY_FILTERED_RANDOM
```
the returned value here doesn't matter. the plan will fail. see the updated docstring. I could have modified the instance resolver to now return errors, but i stuck with a lower impact change.

```
// The passed in node, which contains a replica for the span we are
// attempting to partition, does not satisfy the locality filters, so
// this plan should return an error. In strict mode, the the caller of
// the instance resolver will error if the partition reason is not
// CLOSEST_LOCALITY_MATCH or TARGET HEALTHY. So, to ensure the plan
// fails, return a different reason-- it doesnt matter which one.
```

msbutler replied:
good point, done.

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156474)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kev-cao, @mw5h, and @yuzefovich)

___
*[`pkg/sql/distsql_physical_planner_test.go` line 928 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/156474#-OfXOUAmD_Yn-cxsWj9s:-Ofj4FhE7dDtY-5wLzWg:b84b45a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3719b51051b3cf8276eb24f4d2ae0b0d37db3cd/pkg/sql/distsql_physical_planner_test.go#L928)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I don't understand the beginning of this comment - can you ELI5 :smiley:
</blockquote></details>

heh, it was a silly joke. I personally can't compute this function easily in my head, so i wrote it out.
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1258 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/156474#-OfXQ9S1EtO2SNxwu5ju:-Ofj4bBeDKZiYRbUfhOI:b-kl8g4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3719b51051b3cf8276eb24f4d2ae0b0d37db3cd/pkg/sql/distsql_physical_planner_test.go#L1258)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: n3 doesn't match, right?
</blockquote></details>

good catch!


<!-- Sent from Reviewable.io -->

dt commented on pkg/sql/distsql_physical_planner_test.go:
```diff
@@ -1528,26 +1625,42 @@ func TestPartitionSpans(t *testing.T) {
+
+			// Test strict locality filtering if applicable.
+			if len(locFilters) > 0 {
+				ctx := context.Background()
```
don't you already have a ctx and evalCtx in scope? why shadow with identical redeclarations?

msbutler replied:
i needed to redeclare them to avoid some spanUseAfter finish panics. I've renamed them and added comment to clarify this.

msbutler commented:
TFTR!

bors r=dt
